infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -143,102 +142,147 @@ public void checkpointState() {
-                                          ChannelHandlerContext ctx, IServerRouter r) {
-        TokenRequest req = msg.getPayload();
-        if (req.getNumTokens() == 0) {
-            long max = 0L;
-            boolean hit = false;
-            ImmutableMap.Builder<UUID, Long> streamsLastIssued = ImmutableMap.builder();
-            for (UUID id : req.getStreams()) {
-                lastLocalOffsetMap.compute(id, (k, v) -> {
+    public boolean txnResolution(long timestamp, Set<UUID> streams) {
+        // If the timestamp is -1L, then the transaction automatically commits.
+        if (timestamp != -1L) {
+            AtomicBoolean abort = new AtomicBoolean(false);
+            for (UUID id : streams) {
+                if (abort.get())
+                    break;
+                lastGlobalOffsetMap.compute(id, (k, v) -> {
                     if (v == null) {
-                        streamsLastIssued.put(k, -1L);
                         return null;
+                    } else {
+                        if (v > timestamp) {
+                            log.debug("Rejecting request due to {} > {} on stream {}", v, timestamp, id);
+                            abort.set(true);
+                        }
                     }
-                    streamsLastIssued.put(k, v);
                     return v;
                 });
-                Long lastIssued = lastIssuedMap.get(id);
-                if (lastIssued != null) {
-                    hit = true;
-                }
-                max = Math.max(max, lastIssued == null ? Long.MIN_VALUE : lastIssued);
             }
-            if (!hit) {
-                max = -1L; //no token ever issued
+            if (abort.get()) {
+                return false;
             }
-            if (req.getStreams().size() == 0) {
-                max = globalIndex.get() - 1;
+        }
+        return true;
+    }
+
+    public void returnLatestOffsets(CorfuPayloadMsg<TokenRequest> msg,
+                                    ChannelHandlerContext ctx, IServerRouter r) {
+        TokenRequest req = msg.getPayload();
+        // If no streams are specified in the request, this value returns the last global token issued.
+        long latestGlobalOffset = 0L;
+        boolean hit = false;
+        // Collect the latest local offset for every stream in the request.
+        ImmutableMap.Builder<UUID, Long> requestLatestStreamOffsets = ImmutableMap.builder();
+        for (UUID id : req.getStreams()) {
+            lastLocalOffsetMap.compute(id, (k, v) -> {
+                if (v == null) {
+                    requestLatestStreamOffsets.put(k, -1L);
+                    return null;
+                }
+                requestLatestStreamOffsets.put(k, v);
+                return v;
+            });
+            // Compute the latest global offset across all streams.
+            Long lastIssued = lastGlobalOffsetMap.get(id);
+            if (lastIssued != null) {
+                hit = true;
```

Comment:
add here: 
          latestGlobalOffset = max(latestGlobalOffset, lastIssued); 
and remove the computation in the next line.